doc/source/architecture.rst

```diff
@@ -0,0 +1,213 @@
+.. _ref_architecture:
+
+PyMechanical architecture
+=========================
+
+PyMechanical provides a Python interface to Ansys Mechanical. Mechanical is a
+polyglot [#f1]_ desktop application whose graphical user interface (GUI) runs
+on either the Windows or Linux operating system. Mechanical's APIs are
+implemented in C# using .NET Framework 4.x. They are exposed to both C# and
+two implementations [#f2]_ of Python, IronPython [#f3]_ and CPython.
+
+Within Mechanical, Python scripting enables you to automate repetitive GUI
+actions. This is not unlike other applications developed either by Ansys or
+other software companies. Python scripting leverages the Mechanical API.
+
+Mechanical, like some other applications, is customizable. Using the same
+Mechanical API that you use for scripting, you can implement extensions that
+add to the capabilities of Mechanical. For example, buttons can be added to the
+GUI and custom objects can be added to the data model. Even third-party or
+in-house solvers can be integrated into Mechanical and can take advantage of
+the powerful meshing, generic CAD reader, and the intuitive pre- and post-
+processing experience of Mechanical.
+
+Before discussing how Mechanical's API is implemented, a design pattern known
+as the "Command Pattern" is discussed. This is an abstract pattern that can be
+used in any object oriented system in many programming languages. A high level
+explanation of the pattern (using the Java programming language) can be found
+here: https://howtodoinjava.com/design-patterns/behavioral/command-pattern/.
+A brief explanation follows.
+
+Command pattern
+---------------
+
+Many interactive applications use the Command pattern. In a nutshell, anything
+the user does within the application turns into a Command object, and then the
+Command object is immediately executed. This approach carries some benefits.
+For example, applications that work in this way can choose to implement undo
+and redo by storing an 'inverse' function within the Command and storing a
+stack of executed commands. Undo/Redo is then implemented by walking up and
+down the stack and executing the Command or its inverse function.
+
+More relevant to this discussion is how command patterns can be used to
+implement automation APIs. If every user action is a Command, then that Command
+can serve as an API. In this way, everything the user does can be trivially
+recorded as textual representations of those Commands. These can then be
+replayed using a scripting language supported by the application. In fact, this
+exact command pattern is how the Ansys Electronics Desktop, Ansys Mechanical
+APDL, and the Ansys Workbench applications implement automation APIs for
+scripting.
```

Comment:
```suggestion
More relevant to this discussion is how command patterns can be used to
implement automation APIs. If every user action is a command, then that command
can serve as an API. In this way, everything the user does can be trivially
recorded as textual representations of those commands. These can then be
replayed using a scripting language supported by the application. In fact, this
exact command pattern is how the Ansys Electronics Desktop (AEDT), Ansys Mechanical
APDL, and the Ansys Workbench applications implement automation APIs for
scripting.
```
I've let the "user" here because I think Mohamed wants to distinguish here between a user versus a developer. In general, we do want to use "you" rather than user, but I'll let Mohamed decide here based on his intent.